### DISCUSSION TITLE
Keithley2400_IV Sweep_VIA RS232 - I'd like to increase the size of ':FETCh?'

### DISCUSSION THREAD

**Discussion by kihyeon23** [pymeasure/pymeasure#466] [Q&A]
I'm Kwon, an engineering student. I'm currently producing IV Sweep from keithley2400 products and from Python through rs232. While looking at the manual, I was trying to compensate for various errors and hit a dead end. I'm going to draw a graph with matplotlib, but the number of xvalues and yvalues is not correct. After several attempts, I found that 'yvalues' were fixed to a size of 5. (The graph came out well when each size was adjusted to 5.) The contents of the manual are as follows.
`"You can specify from one to all five elements."`
Please help me to increase the size of ':FETCh?' from 5 so that I can draw a graph that connects the steps I put in. Thank you for reading the long question.
`import sys
    startv = sys.argv[1]
    stopv = sys.argv[2]
    stepv = sys.argv[3]
    filename = sys.argv[4]
    startvprime = float(startv)
    stopvprime = float(stopv)
    stepvprime = float(stepv)
    steps = (stopvprime - startvprime) / stepvprime + 1

    import pyvisa, time
    import serial
    rm = pyvisa.ResourceManager()
    rm.list_resources()
    with rm. open_resource('COM3') as Keithley:
        Keithley.port = 'COM3'
        Keithley.baudrate = 9600
        Keithley.timeout = 25000

    Keithley.open()
    Keithley.read_termination = '\r'
    Keithley.write_termination = '\r'
    Keithley.write("*RST")
    Keithley.write("*IDN?")

    Keithley.write(":SENS:FUNC:CONC OFF")                  
    Keithley.write(":SOUR:FUNC VOLT")                      
    Keithley.write(":SENS:FUNC 'CURR:DC' ")                 

    Keithley.write(":SOUR:VOLT:START ", startv)
    Keithley.write(":SOUR:VOLT:STOP ", stopv)               
    Keithley.write(":SOUR:VOLT:STEP ", stepv)              
    Keithley.write(":SOUR:SWE:RANG AUTO")                   

    Keithley.write(":SENS:CURR:PROT 0.1")                     
    Keithley.write(":SOUR:SWE:SPAC LIN")                    
    Keithley.write(":SOUR:SWE:POIN", str(int(steps)))      
    Keithley.write(":SOUR:SWE:DIR UP")
    Keithley.write(":TRIG:COUN", str(int(steps)))
    Keithley.write(":FORM:ELEM CURR")

    Keithley.write(":SOUR:VOLT:MODE SWE")                   
    Keithley.write(":OUTP ON")                             

    import numpy as np
    result = Keithley.query(":READ?")                       
    yvalues = Keithley.query_ascii_values(":FETCh?")
    Keithley.write(":OUTP OFF")
    Keithley.write(":SOUR:VOLT 0")

    import matplotlib.pyplot as plt
    from matplotlib.animation import FuncAnimation
    from scipy import stats
    xvalues = np.arange(startvprime, stopvprime+1, stepvprime)  
    plt.plot(xvalues, yvalues)
    plt.xlabel(' Drain-Source Voltage (V)')
    plt.ylabel(' Drain-Source Current (mA)')
    plt.title('IV Curve')
    plt.show()
    np.savetxt(filename, (xvalues,yvalues))`

**Comment by kihyeon23** [comment]
error ex) python name.py -10 10 1 savename
=> ValueError: x and y must have same first dimension, but have shapes (21,) and (5,)

**Comment by bilderbuchi** [comment]
I'll convert this into a discussion, as this is not something related to pymeasure (you're not even using pymeasure functionality). Concerning your question, what does the user manual say about FETCH?

**Reply by kihyeon23** [reply]
[:FETCh?]
This query command requests the latest post-processed readings stored in the sample buffer (see Appendix C, “Data Flow”).
After sending this command and addressing the SourceMeter to talk, the readings are sent to the computer. This command does not affect the instrument setup. This command does not trigger source-measure operations; it simply requests the last available readings. Note that this command can repeatedly return the same readings. Until there are new readings, this command continues to return the old readings. For example, assume that the SourceMeter performed 20 source-measure operations. The :FETCh? command will request the readings for those 20 source-measure operations. If :FETCh? is sent while performing source measure operations (ARM annunciator on), it will not be executed until the SourceMeter goes back into idle.
The readings that are acquired depend on which data elements are selected (Section 18, “Data elements”), and what the instrument is presently programmed to source-measure. Measure readings take priority over source readings, and functions not sourced or measured are assigned the NAN (not a number) value of +9.91e37.
For example, assume that voltage, current and resistance readings are selected as data elements, and the instrument is programmed to Source V and Measure I. A reading string acquired by :FETch? will include the programmed V-Source value and the I-Measure reading. The reading for resistance will be NAN since resistance was not measured.
If the SourceMeter is instead programmed to Source V and Measure V, the voltage reading will be the V-Measure reading (not the programmed V-Source value). Both current and resistance readings will be NANs (current is not measured or sourced).

[NOTE]
The :FETCh? command is automatically asserted when the :READ? or :MEASure? command is sent.

**Reply by bilderbuchi** [reply]
Maybe you'll have to adjust the buffer size, although this text seems to suggest that that happens automatically? 
Also, try printing the content of `yvalues` to make sure the values are in the format you expect.
